### PR TITLE
Translate to zh-hans: How Are Function Components Different from Classes?

### PR DESCRIPTION
**Pull request by phyzess**
This pull request contains a translation of the article "How Are Function Components Different from Classes?" to zh-hans.

### CODE REVIEW

**Comment by gaearon** [comment]
There's an earlier PR in https://github.com/gaearon/overreacted.io/pull/409, can you please coordinate with its author?

**Comment by MmmmyPh** [comment]
My fault. I forgot to search first. But how should I do that? I'm not very familiar with pr.

**Comment by Jemair** [comment]
@MmmmyPh did a great job, please ACK this pr directly. I'll close mine later.

thanks @gaearon 

PS: 

Before you translate a post, you may take a look at the pr list whether there has been one in progress.

If not, send a pull request with a `WIP`(working in progress) tag in case duplication of effort.

And remove the tag after you knocked off.

**Comment by MmmmyPh** [comment]
Thanks @Jemair. I'll pay attention to not making this mistake again.

File: src/pages/how-are-function-components-different-from-classes/index.zh-hans.md [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+
+函数捕获了他们的props和state —— 因此它们的标识也同样重要。这不是一个bug，而是一个函数式组件的特性。例如，对于`useEffect`或者`useCallback`来说，函数不应该被排除在“依赖数组”之外。（正确的解决方案通常是使用上面说过的`useReducer`或者`useRef` —— 我们将很快会在文档中说明如何在它们两者中进行选择。）
+
+当我们用函数来编写大部分的React代码时，我们需要调整关于[优化代码](https://github.com/ryardley/hooks-perf-issues/pull/3)和[什么变量会随着时间改变](https://github.com/facebook/react/issues/14920)的认知与直觉。
+
+正如[Fredrik所说](https://mobile.twitter.com/EphemeralCircle/status/1099095063223812096)：
+
+>到目前为止，我发现的有关于hooks的最好的心里规则是“写代码时要认为任何值都可以随时更改”。
+
+函数也不例外。这需要一段时间才能成为React学习资料中的常识。它需要一些从类的思维方式中进行一些调整。但我希望这篇文章能够帮助你以新的眼光来看待它。
```
Comment (tonghuashuo) [inline src/pages/how-are-function-components-different-from-classes/index.zh-hans.md]: 需要在基于 class 的思维方式上做一些调整。

- Optimize translation.

**Comment by gaearon** [comment]
I'm gonna get this in, please feel free to send follow-up fixes!

**Pull request merged**